format/neko/Reader.hx

```diff
@@ -1,163 +1,310 @@
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+package format.neko;
+import format.neko.Data;
+
+class Reader {
+
+	var i : haxe.io.Input;
+
+	public function new( i : haxe.io.Input ) {
+		this.i = i;
+	}
+
+	function error() : Dynamic {
+		throw "Invalid file format";
+		return null;
+	}
+
+	inline function readInt() {
+		#if haxe3
+		return i.readInt32();
+		#else
+		return i.readUInt30();
+		#end
+	}
+
+	function readDebugInfos() : DebugInfos {
+		var nfiles = i.readByte();
+		var manyFiles = false;
+		if( nfiles >= 0x80 ) {
+			nfiles = ((nfiles & 0x7F) << 8) | i.readByte();
+			manyFiles = true;
+		}
+		var files = [];
+		for( k in 0...nfiles )
+			files[k] = i.readUntil(0);
+		var npos = readInt();
+		var curfile = files[0];
+		var curline = 0;
+		var curpos = null;
+		var p = 0;
+		var pos = alloc(npos);
+		while( p < npos ) {
+			var c = i.readByte();
+			if( c & 1 != 0 ) {
+				c >>= 1;
+				if( manyFiles )
+					c = (c << 8) | i.readByte();
+				curfile = files[c];
+				curpos = null;
+			} else if( c & 2 != 0 ) {
+				var delta = c >> 6;
+				var count = (c >> 2) & 15;
+				if( curpos == null )
+					curpos = { file : curfile, line : curline };
+				for( k in 0...count )
+					pos[p++] = curpos;
+				if( delta != 0 ) {
+					curline += delta;
+					curpos = null;
+				}
+			} else if( c & 4 != 0 ) {
+				curline += c >> 3;
+				curpos = { file : curfile, line : curline };
+				pos[p++] = curpos;
+			} else {
+				var b1 = i.readByte();
+				var b2 = i.readByte();
+				curline = (c >> 3) | (b1 << 5) | (b2 << 13);
+				curpos = { file : curfile, line : curline };
+				pos[p++] = curpos;
+			}
+		}
+		return pos;
+	}
+
+	function alloc<T>( size : Int ) : Array<T> {
+		var a = new Array<T>();
+		if( size > 0 ) a[size-1] = null;
+		return a;
+	}
+
+	function hash( s : String ) : Int {
+		var h = 0;
+		#if neko
+		for( i in 0...s.length )
+			h = 223 * h + s.charCodeAt(i);
+		return h;
+		#else
+		for( i in 0...s.length )
+			h = (((223 * h) >> 1) + s.charCodeAt(i)) << 1;
+		return h >> 1;
+		#end
+	}
+
+	public function read() : Data {
+		if( i.readString(4) != "NEKO" ) error();
+		var nglobals = readInt();
+		var nfields = readInt();
+		var codesize = readInt();
+		if (nglobals < 0 || nglobals > 0xffff
+				|| nfields < 0 || nfields > 0xffff
+				|| codesize < 0 || codesize > 0xffffff)
+				error();
+
+		// globals
+		var globals = alloc(nglobals);
+		for( k in 0...nglobals )
+			globals[k] = switch( i.readByte() ) {
+			case 1:
+				GlobalVar(i.readUntil(0));
+			case 2:
+				var pos = i.readUInt24();
+				var nargs = i.readByte();
+				GlobalFunction(pos, nargs);
+			case 3:
+				GlobalString(i.readString(i.readUInt16()));
+			case 4:
+				GlobalFloat(i.readUntil(0));
+			case 5:
+				GlobalDebug(readDebugInfos());
+			case 6:
+				GlobalVersion(i.readByte());
+			default:
+				error();
+			}
+
+		// fields
+		var fields = alloc(nfields);
+		var fieldHashes : Map<Int, String> = [];
```

Comment:
For Haxe 3 compatibility, this should be `var fieldHashes = new Map<Int, String>();`